feat(plan): add plan sub-directory support

flotter opened the pull request:
Pebble currently only supports loading configuration layer files from the root of $PEBBLE/layers.

Add support for adding sub-directory support at the root level (2 levels in total only). [Spec KO071.](https://docs.google.com/document/d/1-9GLCJTx9o_hSd0JGvDp7o5xNQjkJMvLam6o1Ba_bYM/edit?tab=t.0)

This scheme has an impact on the meaning of ```label``` and ```order```. See the following examples to understand the new mapping:

- $PEBBLE/layers/001-foo.yaml
Order: 001-000 => 1000
Label: foo

- $PEBBLE/layers/005-bar.d/010-abc.yaml
Order: 005-010 => 5010
Label: bar/abc

- $PEBBLE/layers/005-bar.d/012-def.yaml
Order: 005-012 => 5012
Label: bar/def

- $PEBBLE/layers/006-baz.yaml
Order: 006-000 => 6000
Label: baz

Directory support allows other uses of Pebble using read-only file systems more flexibility by now giving the option to bind-mount a sub-directory to a writable disk location, for example.

flotter commented:
@benhoyt Hi Ben, would it be possible for you to give this a initial scan for me please, when you have some time? I want to see if there are early course corrections needed.

Outstanding tasks:

1. Unit Tests outside of plan still needs fixing
2. Unit Tests missing for new functionality in Plan (current tests pass).
3. Some logic required in planstate (Append / Combine) to deal with requests coming in over the API.

benhoyt reviewed (commented):
I've had a cursory look. Overall structure looks reasonable.

My main comment is the regexes are getting pretty hard to understand (and get right!). I wonder if we can have a simple regex that just matches the "001-label" part, and do the other parts with `strings.Split` and `filepath.Ext` / `filepath.Base`.

The other thing I wonder if you should consider is combining the directory traversal with the parsing -- because when you're traversing the directory you've already split on `/` so you have more information. I'm not sure, but it seems like that would eliminate a bit of the parsing by design.

flotter commented on client/plan.go:
```diff
@@ -25,6 +25,11 @@ type AddLayerOptions struct {
 	Combine bool
 
+	// Inner true means a new layer append may go into an existing
+	// subdirectory, even through it may not result in appending it
```
Fixed, thank you. Resolving.

flotter commented:
> I've had a cursory look. Overall structure looks reasonable.
> 
> My main comment is the regexes are getting pretty hard to understand (and get right!). I wonder if we can have a simple regex that just matches the "001-label" part, and do the other parts with `strings.Split` and `filepath.Ext` / `filepath.Base`.
> 
> The other thing I wonder if you should consider is combining the directory traversal with the parsing -- because when you're traversing the directory you've already split on `/` so you have more information. I'm not sure, but it seems like that would eliminate a bit of the parsing by design.

@benhoyt thank you for finding the time to have a look. Exactly the guidance I needed thank you. Your proposal makes a lot of sense and I made the following changes in line with what you are proposing:

1. I removed all the complex regex expressions and kept only one simple one (similar to the original one).
2. I moved the loading into the tree traversal code, which meant I have access to the already split orders/labels.

This allowed me to remove and simplify lots of complex code.

TODO: I will proceed with completing the outstanding tasks here, and report back for a more thorough review.

benhoyt commented on internals/overlord/planstate/manager.go:
```diff
@@ -155,19 +162,86 @@ func (m *PlanManager) CombineLayer(layer *plan.Layer) error {
+		// an order?
+		newSubLabel, _, hasSub := strings.Cut(newLayer.Label, "/")
+		lastIndex := layersCount - 1
+		for i := range m.plan.Layers {
```
This seems like an unintuitive way to iterate through a slice backwards (by looping the index forward). Can we loop `i` backwards instead?

```go
for i := len(m.plan.Layers) - 1; i >= 0; i-- {
    layer := m.plan.Layers[i]
    // ALSO BELOW: newIndex = i + 1
}
```

flotter replied:
Improved, thanks. Much cleaner. Resolving.

benhoyt reviewed (commented):
Looks very good, thanks! Some comments, but almost all minor / stylistic.

anpep commented on internals/plan/plan.go:
```diff
@@ -1349,44 +1393,169 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
+// │ 002-foo.d              │ 002       │ foo       │ .d    │
+// │                        │           │           │       │
+// └────────────────────────┴───────────┴───────────┴───────┘
+var configEntryExp = regexp.MustCompile(`^([0-9]{3})-([a-z](?:-?[a-z0-9]){2,})(.yaml|.d)$`)
```
There are multiple occurrences of this, don't forget to apply elsewhere!

flotter replied:
Changed in 6 places from Exp -> Regexp. Resolving.

anpep commented on client/plan.go:
```diff
@@ -25,6 +25,11 @@ type AddLayerOptions struct {
 	// has the given label. False (the default) means append a new layer.
 	Combine bool
 
+	// Inner true means a new layer append may go into an existing
```
```suggestion
	// Inner, when true, means a new layer append may go into an existing
```

flotter replied:
Improved, thank you. Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -154,6 +154,28 @@ func (p *Plan) MarshalYAML() (interface{}, error) {
+// number. Each layer configuration also has a unique label, used for locating
+// and updating a specific layer configuration.
+//
+// Pebble supports a two level layer configuration directory structure. In the
```
Fixed, thank you.

flotter commented on internals/plan/plan.go:
```diff
@@ -154,6 +154,28 @@ func (p *Plan) MarshalYAML() (interface{}, error) {
+//
+// Pebble supports a two level layer configuration directory structure. In the
+// root layers directory, both layer files and layer sub-directories are
+// allowed. Within a sub-directory, only layer files.
```
Doh! Fixed. Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -154,6 +154,28 @@ func (p *Plan) MarshalYAML() (interface{}, error) {
+// │                            │                 │         │
+// │ 003-baz.yaml               │ 003-000 => 3000 │ baz     │
+// │                            │                 │         │
+// └────────────────────────────┴─────────────────┴─────────┘
```
Added a reference to ReadLayersDir, and simplified the remaining tables to the simpler text format. Thanks! Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -1349,44 +1393,169 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
 }
 
-var fnameExp = regexp.MustCompile("^([0-9]{3})-([a-z](?:-?[a-z0-9]){2,}).yaml$")
+// ReadLayersDir loads the YAML layer files from the first two directory
+// levels starting at layersDir in the order as specified by the order
+// directory and file prefixes. Note that the directory and file suffix
+// is dropped for the label.
```
Fixed. Thank you. Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -1349,44 +1393,169 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
+	return layer, nil
+}
+
+// configEntryExp matches either a valid config layer YAML file name or it
+// matches a valid config layer directory.
```
Fixed, Thank you. Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -1349,44 +1393,169 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
+// root directory, in which case l2Entry must be nil. If the file is
+// inside a sub-directory, l1Entry must supply information on the directory,
+// while l2Entry information on the file name itself.
+func configLayerLoad(layersDir string, l1Entry *configEntry, l2Entry *configEntry) (*Layer, error) {
```
Changed in 4 places. Thanks. Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -1227,7 +1249,29 @@ func (p *Plan) checkCycles() error {
 func ParseLayer(order int, label string, data []byte) (*Layer, error) {
+	// This function can be called directly over the daemon API. We
+	// must fail the API request if the label is not valid.
+	match := labelExp.FindStringSubmatch(label)
```
Resolving.

flotter commented on internals/plan/plan.go:
```diff
@@ -1349,44 +1393,169 @@ func validServiceAction(action ServiceAction, additionalValid ...ServiceAction)
+	path := filepath.Join(layersDir, l1Entry.name)
+	if l2Entry != nil {
+		// Config layer is inside a sub-directory.
+		label = filepath.Join(label, l2Entry.label)
```
You are right. I prefer to fix it as per your proposal. Fixed. Resolving.

flotter commented on internals/plan/plan_test.go:
```diff
@@ -2156,3 +2157,164 @@ func (s *S) TestSectionOrder(c *C) {
+	path := filepath.Join(base, name)
+	err := os.MkdirAll(filepath.Dir(path), 0777)
+	c.Assert(err, IsNil)
+	file, err := os.OpenFile(path, os.O_RDONLY|os.O_CREATE, 0644)
```
Yes, I think you highlight an important point, thank you. I added a simple service to each layer, with a command specific to the layer name, so I can verify the content match the expected layer. So I consider this fixed. Resolving.

flotter commented on internals/overlord/planstate/manager_test.go:
```diff
@@ -508,3 +509,82 @@ services:
+
+func (ps *planSuite) TestAppendLayersNoInner(c *C) {
+	plan.RegisterSectionExtension(testField, testExtension{})
+	defer plan.UnregisterSectionExtension(testField)
```
The append and combine logic, and the underlying parser logic can be impacted by having an extension registered. Since the content of the file here is none (empty), it may not be strictly necessary, but it does perform the test using the most complex configuration, i.e. with an extension active. I think this is OK, as it mimics the other combine and append unit tests, but I am happy to remove it if you prefer ?

benhoyt replied:
No, that's fine -- I just wanted to know the reason.

flotter commented on internals/overlord/planstate/manager_test.go:
```diff
@@ -508,3 +509,82 @@ services:
+	c.Assert(err, ErrorMatches, ".*cannot insert sub-directory.*")
+}
+
+func (ps *planSuite) TestAppendLayersOrderAllocation(c *C) {
```
Changed both to:

```
TestAppendLayersWithInner
TestAppendLayersWithoutInner
```

flotter replied:
Resolving.

flotter commented on internals/overlord/planstate/manager.go:
```diff
@@ -155,19 +162,86 @@ func (m *PlanManager) CombineLayer(layer *plan.Layer) error {
+// The new incoming layer only supplies the label, which may include
+// a sub-directory prefix. Normally without a sub-directory prefix,
+// the new layer will always be appended, which means incrementing
+// the root level order (+ 1000). If a sub-directory already exist
```
Fixed, Thank you. Resolving.

flotter commented on internals/daemon/api_plan.go:
```diff
@@ -43,6 +43,7 @@ func v1PostLayers(c *Command, r *http.Request, _ *UserState) Response {
 	var payload struct {
 		Action  string `json:"action"`
 		Combine bool   `json:"combine"`
+		Inner   bool   `json:"inner"`
```
Added, thank you. Resolving.

benhoyt reviewed (approved):
Thanks!
